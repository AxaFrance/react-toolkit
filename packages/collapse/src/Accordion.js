import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { ClassManager, Constants } from '@axa-fr/react-toolkit-core';

const propTypes = {
  ...Constants.propTypes,
  children: PropTypes.node.isRequired,
  defaultCollapse: PropTypes.bool,
  onlyOne: PropTypes.bool,
};
const defaultClassName = 'af-accordion';
const defaultProps = {
  ...Constants.defaultProps,
  defaultCollapse: true,
  onlyOne: true,
  className: defaultClassName,
};

export const AccordionBase = props => {
  const { children, className, classModifier, handleToggle, collapses } = props;
  const renderedChildren = React.Children.map(children, child => {
    const index = Array.isArray(children) ? children.indexOf(child) : 0;
    return React.cloneElement(child, {
      ...child.props,
      index,
      onToggle: handleToggle,
      collapse:
        collapses[index] !== undefined
          ? collapses[index]
          : child.props.collapse,
      className: child.props.className,
      classModifier: child.props.classModifier,
    });
  });
  const componentClassName = ClassManager.getComponentClassName(
    className,
    classModifier,
    defaultClassName
  );
  return (
    <div
      className={componentClassName}
      role="tablist"
      aria-multiselectable="true">
      {renderedChildren}
    </div>
  );
};

export const handleToggle = (state, { onlyOne, children }) => e => {
  const newState = { collapses: [...state.collapses] };
  if (e.collapse && onlyOne) {
    if (Array.isArray(children)) {
      children.forEach((child, index) => {
        newState.collapses[index] = false;
      }, this);
    } else if (children) {
      newState.collapses[0] = false;
    }
  }
  newState.collapses[e.index] = e.collapse;
  return newState;
};

AccordionBase.propTypes = propTypes;
AccordionBase.defaultProps = defaultProps;

const EnhancedComponent = (props) => {
  const [state, setState] = useState({ collapses: [] });

  const toggle = (e) => {
    const newState = handleToggle(state, props)(e);
    setState(newState);
  };

  return (<AccordionBase {...props} {...state} handleToggle={toggle} />);
};

EnhancedComponent.displayName = 'Accordion';

export default EnhancedComponent;
