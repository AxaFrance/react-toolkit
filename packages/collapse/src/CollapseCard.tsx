import React from 'react';

import {
  Constants,
  InputManager,
  ClassManager,
} from '@axa-fr/react-toolkit-core';
import Body, { BodyProps } from './Body';
import Header, { HeaderProps, HeaderToggleElement } from './Header';

const defaultClassName = 'af-accordion__item';

const defaultProps = {
  ...Constants.defaultProps,
  className: defaultClassName,
  index: 0,
  isOpen: false,
};

export type CollapseProps = Partial<typeof defaultProps> & {
  id?: string;
  children?: React.ReactNode[];
  onToggle?: (element: HeaderToggleElement) => void;
};

const CollapseCard = ({
  id,
  className,
  classModifier,
  isOpen,
  children,
  index,
  onToggle,
}: CollapseProps) => {
  const [headerId, setHeaderId] = React.useState(InputManager.getInputId(id));
  const [open, setOpen] = React.useState(isOpen);

  React.useEffect(() => {
    setHeaderId(InputManager.getInputId(id));
  }, [id]);

  React.useEffect(() => {
    setOpen(isOpen);
  }, [isOpen]);

  const handleToggle = (e: HeaderToggleElement) => {
    onToggle && onToggle(e);
    setOpen(!open);
  };

  const renderChild = (child: React.ReactElement<HeaderProps | BodyProps>) => {
    const bodyId = `${headerId}body`;
    switch (child.type) {
      case Header:
        return React.cloneElement(child, {
          isOpen: open,
          onToggle: (e: HeaderToggleElement) => handleToggle(e),
          classModifier,
          id: headerId,
          index,
          href: `#${bodyId}`,
          ariaControls: bodyId,
          key: headerId,
        });
      case Body:
        return React.cloneElement(child, {
          isOpen: open,
          classModifier,
          ariaLabelledby: headerId,
          id: bodyId,
          key: bodyId,
        });
      default:
        return child;
    }
  };

  let newClassModifier = open ? 'open' : '';
  newClassModifier = classModifier
    ? `${newClassModifier} ${classModifier}`
    : `${newClassModifier}`;

  const componentClassName = ClassManager.getComponentClassName(
    className,
    newClassModifier,
    defaultClassName
  );

  return <div className={componentClassName}>{children.map(renderChild)}</div>;
};

CollapseCard.defaultProps = defaultProps;
CollapseCard.Body = Body;
CollapseCard.Header = Header;

export default CollapseCard;
