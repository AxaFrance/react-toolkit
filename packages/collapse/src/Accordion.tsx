import React from 'react';
import { ClassManager, Constants } from '@axa-fr/react-toolkit-core';
import { CollapseProps } from './CollapseCard';
import { HeaderToggleElement } from './Header';

const defaultClassName = 'af-accordion';

const defaultProps = {
  ...Constants.defaultProps,
  onlyOne: false,
  className: defaultClassName,
};

type AccordionProps = {
  className: string;
  classModifier: string;
  children: React.ReactElement<CollapseProps>[];
  handleToggle: (e: HeaderToggleElement) => void;
  collapses: boolean[];
};

const Accordion = ({
  className,
  classModifier,
  children,
  handleToggle,
  collapses,
}: AccordionProps) => {
  const renderedChildren = React.Children.map(children, (child, index) => {
    let mixCallback = handleToggle;
    if (child.props.onToggle) {
      mixCallback = (e: HeaderToggleElement) => {
        handleToggle(e);
        child.props.onToggle(e);
      };
    }
    return React.cloneElement(child, {
      ...child.props,
      index,
      onToggle: mixCallback,
      isOpen:
        collapses[index] !== undefined ? collapses[index] : child.props.isOpen,
      className: child.props.className,
      classModifier: child.props.classModifier,
    });
  });
  const componentClassName = ClassManager.getComponentClassName(
    className,
    classModifier,
    defaultClassName
  );
  return (
    <div
      className={componentClassName}
      role="tablist"
      aria-multiselectable="true">
      {renderedChildren}
    </div>
  );
};

Accordion.defaultProps = defaultProps;

export const handleToggleState = (
  collapses: boolean[],
  { onlyOne, children }: EnhancedProps
) => (e: HeaderToggleElement): boolean[] => {
  if (e.isOpen && onlyOne) {
    if (Array.isArray(children)) {
      return children.map((_child, index) =>
        e.index === index ? e.isOpen : false
      );
    }
    return [e.isOpen];
  }
  const newCollapses = [...collapses];
  newCollapses[e.index] = e.isOpen;
  return newCollapses;
};

export type EnhancedProps = Partial<typeof defaultProps> & {
  onlyOne?: boolean;
  className?: string;
  classModifier?: string;
  children: React.ReactElement<CollapseProps>[];
};

const EnhancedComponent = (props: EnhancedProps) => {
  const [collapses, setCollapses] = React.useState<boolean[]>([]);

  const toggle = (e: HeaderToggleElement) => {
    const newCollapses = handleToggleState(collapses, props)(e);
    setCollapses(newCollapses);
  };

  return <Accordion {...props} collapses={collapses} handleToggle={toggle} />;
};

EnhancedComponent.displayName = 'Accordion';

export default EnhancedComponent;
