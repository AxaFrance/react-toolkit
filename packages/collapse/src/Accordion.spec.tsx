import React from 'react';
import { fireEvent, render, waitFor, screen } from '@testing-library/react';
import Accordion, { handleToggleState, EnhancedProps } from './Accordion';
import CollapseCard from './CollapseCard';

const setup = (props: Omit<EnhancedProps, 'children'>) => {
  const onToggle = jest.fn();
  const utils = render(
    <Accordion {...props}>
      <CollapseCard id="one" isOpen>
        <CollapseCard.Header>Header 1</CollapseCard.Header>
        <CollapseCard.Body>Text 1</CollapseCard.Body>
      </CollapseCard>
      <CollapseCard id="two" isOpen={false}>
        <CollapseCard.Header>Header 2</CollapseCard.Header>
        <CollapseCard.Body>Text 2</CollapseCard.Body>
      </CollapseCard>
      <CollapseCard id="three" isOpen={false} onToggle={onToggle}>
        <CollapseCard.Header>Header 3</CollapseCard.Header>
        <CollapseCard.Body>Text 3</CollapseCard.Body>
      </CollapseCard>
    </Accordion>
  );
  return { ...utils };
};

describe('Accordion', () => {
  it('renders Accordion correctly', () => {
    const { asFragment } = setup({});
    expect(asFragment()).toMatchSnapshot();
  });

  it('should open only one collapse (onlyOne)', async () => {
    const { container } = setup({ onlyOne: true });

    fireEvent.click(screen.getByText('Header 3'));

    await waitFor(() => {
      expect(
        container.querySelectorAll('.af-accordion__collapse--open').length
      ).toBe(1);
      expect(
        container
          .querySelector('.af-accordion__item--open')
          .querySelector('.af-accordion__block').innerHTML
      ).toEqual('Text 3');
    });
  });

  it('handleToggle should return correct value', () => {
    let oldCollapses = [false];
    let props = { onlyOne: true, children: {} };
    let event = { isOpen: true, index: 0, id: 'id' };
    // @ts-ignore
    let collapses = handleToggleState(oldCollapses, props)(event);
    expect(collapses).toEqual([true]);

    oldCollapses = [false, true, false];
    props = { onlyOne: true, children: [{}, {}, {}] };
    event = { isOpen: true, index: 0, id: 'id' };
    // @ts-ignore
    collapses = handleToggleState(oldCollapses, props)(event);
    expect(collapses).toEqual([true, false, false]);

    event = { isOpen: false, index: 1, id: 'id2' };
    // @ts-ignore
    collapses = handleToggleState(oldCollapses, props)(event);
    expect(collapses).toEqual([false, false, false]);

    event = { isOpen: true, index: 0, id: 'id' };
    props = { onlyOne: false, children: [{}, {}, {}] };
    // @ts-ignore
    collapses = handleToggleState(oldCollapses, props)(event);
    expect(collapses).toEqual([true, true, false]);
  });
});
