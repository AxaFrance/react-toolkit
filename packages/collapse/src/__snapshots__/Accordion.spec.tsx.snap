// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion renders Accordion correctly 1`] = `
<DocumentFragment>
  <div
    aria-multiselectable="true"
    class="af-accordion"
    role="tablist"
  >
    <div
      class="af-accordion__item af-accordion__item--open"
    >
      <div
        class="af-accordion__item-header"
        id="one"
        role="tab"
      >
        <a
          aria-controls="onebody"
          aria-expanded="true"
          class="af-accordion__item-toggle"
          data-toggle="isOpen"
          href="#onebody"
        >
          <h3
            class="af-accordion__item-title"
          >
            <span
              class="glyphicon glyphicon-menu-up"
            />
            Header 1
          </h3>
        </a>
      </div>
      <div
        aria-labelledby="one"
        class="af-accordion__collapse af-accordion__collapse--open"
        id="onebody"
        role="tabpanel"
        style="max-height: 0px;"
      >
        <div
          class="af-accordion__block"
        >
          Text 1
        </div>
      </div>
    </div>
    <div
      class="af-accordion__item"
    >
      <div
        class="af-accordion__item-header"
        id="two"
        role="tab"
      >
        <a
          aria-controls="twobody"
          aria-expanded="false"
          class="af-accordion__item-toggle"
          data-toggle="isOpen"
          href="#twobody"
        >
          <h3
            class="af-accordion__item-title"
          >
            <span
              class="glyphicon glyphicon-menu-down"
            />
            Header 2
          </h3>
        </a>
      </div>
      <div
        aria-labelledby="two"
        class="af-accordion__collapse af-accordion__collapse--close"
        id="twobody"
        role="tabpanel"
        style="max-height: 0px;"
      >
        <div
          class="af-accordion__block"
        >
          Text 2
        </div>
      </div>
    </div>
    <div
      class="af-accordion__item"
    >
      <div
        class="af-accordion__item-header"
        id="three"
        role="tab"
      >
        <a
          aria-controls="threebody"
          aria-expanded="false"
          class="af-accordion__item-toggle"
          data-toggle="isOpen"
          href="#threebody"
        >
          <h3
            class="af-accordion__item-title"
          >
            <span
              class="glyphicon glyphicon-menu-down"
            />
            Header 3
          </h3>
        </a>
      </div>
      <div
        aria-labelledby="three"
        class="af-accordion__collapse af-accordion__collapse--close"
        id="threebody"
        role="tabpanel"
        style="max-height: 0px;"
      >
        <div
          class="af-accordion__block"
        >
          Text 3
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
